frontend_tests/puppeteer_lib/common.js

```diff
@@ -120,6 +125,167 @@ class CommonUtils {
+                actually acks the message, the message will
+                have a temporary id and will not have all
+                the normal message controls.
+                For the Casper tests, we want to avoid all
+                the edge cases with locally echoed messages.
+                In order to make sure a message is processed,
+                we use internals to determine the following:
+                    - has message_list even been updated with
+                      the message with out content?
+                    - has the locally_echoed flag been cleared?
+                But for the final steps we look at the
+                actual DOM (via JQuery):
+                    - is it visible?
+                    - does it look to have been
+                      re-rendered based on server info?
+            */
+            const last_msg = current_msg_list.last();
+
+            if (last_msg.raw_content !== content) {
+                return false;
+            }
+
+            if (last_msg.locally_echoed) {
+                return false;
+            }
+
+            const row = rows.last_visible();
+
+            if (rows.id(row) !== last_msg.id) {
+                return false;
+            }
+
+            /*
+                Make sure the message is completely
+                re-rendered from its original "local echo"
+                version by looking for the star icon.  We
+                don't add the star icon until the server
+                responds.
+            */
+            return row.find('.star').length === 1;
+        }, content);
+    }
+
+    // Wait for any previous send to finish, then send a message.
+    async send_message(page, type, params) {
+        // If a message is outside the view, we will skip
+        // validation later.
+        const outside_view = params.outside_view;
+        delete params.outside_view;
+
+        await page.waitForSelector('#compose-send-button:enabled');
+        await page.waitForSelector('#compose-textarea');
+
+        if (type === "stream") {
+            await page.keyboard.press('KeyC');
+        } else if (type === "private") {
+            await page.keyboard.press("KeyX");
+            const recipients = params.recipient.split(', ');
+            for (let i = 0; i < recipients.length; i += 1) {
+                await this.set_pm_recipient(page, recipients[i]);
+            }
+            delete params.recipient;
+        } else {
+            assert.fail("`send_message` got invalid message type");
+        }
+
+        if (params.stream) {
+            params.stream_message_recipient_stream = params.stream;
+            delete params.stream;
+        }
+
+        if (params.subject) {
+            params.stream_message_recipient_topic = params.subject;
+            delete params.subject;
+        }
+
+        await this.fill_form(page, 'form[action^="/json/messages"]', params);
+        await this.turn_off_press_enter_to_send(page);
```

Comment:
Let's call this `ensure_enter_does_not_send`?  Since it only does a toggle if it's necessary.